File: event-service/src/test/java/com/github/manosbatsis/eventservice/service/UserEventServiceImplTest.java
package com.github.manosbatsis.eventservice.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.BDDMockito.given;

import com.github.manosbatsis.eventservice.model.UserEvent;
import com.github.manosbatsis.eventservice.model.UserEventKey;
import com.github.manosbatsis.eventservice.repository.UserEventRepository;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.junit.jupiter.SpringExtension;

@ExtendWith(SpringExtension.class)
@Import(UserEventServiceImpl.class)
class UserEventServiceImplTest {

  @Autowired private UserEventService userEventService;

  @MockBean private UserEventRepository userEventRepository;

  @Test
  void testGetUserEventsWhenThereIsNone() {
    given(userEventRepository.findByKeyUserId(anyLong())).willReturn(Collections.emptyList());

    List<UserEvent> userEvents = userEventService.getUserEvents(1L);

    assertThat(userEvents).isNotNull();
    assertThat(userEvents).isEmpty();
  }

  @Test
  void testGetUserEventsWhenThereIsOne() {
    UserEvent userEvent = getDefaultUserEvent();
    given(userEventRepository.findByKeyUserId(anyLong()))
        .willReturn(Collections.singletonList(userEvent));

    List<UserEvent> userEvents = userEventService.getUserEvents(1L);

    assertThat(userEvents).isNotNull();
    assertThat(userEvents.size()).isEqualTo(1);
    assertThat(userEvents.get(0)).isEqualTo(userEvent);
  }

  @Test
  void testSaveUserEvent() {
    UserEvent userEvent = getDefaultUserEvent();
    given(userEventRepository.save(any(UserEvent.class))).willReturn(userEvent);

    UserEvent userEventSaved = userEventService.saveUserEvent(userEvent);

    assertThat(userEventSaved).isNotNull();
    assertThat(userEventSaved).isEqualTo(userEvent);
  }

  private UserEvent getDefaultUserEvent() {
    return new UserEvent(new UserEventKey(1L, new Date()), "type", "data");
  }
}

File: event-service/src/test/kotlin/com/github/manosbatsis/services/event/service/UserEventServiceImplTest.kt
package com.github.manosbatsis.services.event.service

import com.github.manosbatsis.services.event.model.UserEvent
import com.github.manosbatsis.services.event.model.UserEventKey
import com.github.manosbatsis.services.event.repository.UserEventRepository
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentMatchers
import org.mockito.BDDMockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.context.annotation.Import
import org.springframework.test.context.junit.jupiter.SpringExtension
import java.util.Date

@ExtendWith(SpringExtension::class)
@Import(UserEventServiceImpl::class)
class UserEventServiceImplTest {
    @Autowired
    private lateinit var userEventService: UserEventService

    @MockBean
    private lateinit var userEventRepository: UserEventRepository

    @Test
    fun testGetUserEventsWhenThereIsNone() {
        BDDMockito.given(
            userEventRepository.findByKeyUserId(ArgumentMatchers.anyLong()),
        ).willReturn(emptyList())
        val userEvents = userEventService.findByUserId(1L)
        Assertions.assertThat(userEvents).isNotNull()
        Assertions.assertThat(userEvents).isEmpty()
    }

    @Test
    fun testGetUserEventsWhenThereIsOne() {
        val userEvent = defaultUserEvent
        BDDMockito.given(
            userEventRepository.findByKeyUserId(ArgumentMatchers.anyLong()),
        )
            .willReturn(listOf(userEvent))
        val userEvents = userEventService.findByUserId(1L)
        Assertions.assertThat(userEvents).isNotNull()
        Assertions.assertThat(userEvents.size).isEqualTo(1)
        Assertions.assertThat(userEvents[0]).isEqualTo(userEvent)
    }

    @Test
    fun testSaveUserEvent() {
        val userEvent = defaultUserEvent
        BDDMockito.given(
            userEventRepository.save(
                ArgumentMatchers.any(
                    UserEvent::class.java,
                ),
            ),
        ).willReturn(userEvent)
        val userEventSaved = userEventService.save(userEvent)
        Assertions.assertThat(userEventSaved).isNotNull()
        Assertions.assertThat(userEventSaved).isEqualTo(userEvent)
    }

    private val defaultUserEvent: UserEvent
        private get() = UserEvent(UserEventKey(1L, Date()), "type", "data")
}
